bootloader/lkenv/lkenv.go

```diff
@@ -287,21 +336,120 @@ func (l *Env) GetBootPartition(kernel string) (string, error) {
 // returned, otherwise the reference is removed and nil is returned.
 // Note that to persist this change the env must be saved afterwards with Save.
 func (l *Env) RemoveKernelRevisionFromBootPartition(kernel string) error {
-	for x := range l.env.Bootimg_matrix {
-		if "" != cToGoString(l.env.Bootimg_matrix[x][MATRIX_ROW_PARTITION][:]) {
-			if kernel == cToGoString(l.env.Bootimg_matrix[x][MATRIX_ROW_KERNEL][:]) {
-				l.env.Bootimg_matrix[x][1][MATRIX_ROW_PARTITION] = 0
-				return nil
-			}
-		}
-	}
-	return fmt.Errorf("cannot find defined [%s] boot image partition", kernel)
+	return l.variant.removeKernelFromBootPart(kernel)
 }
 
 // GetBootImageName return expected boot image file name in kernel snap
 func (l *Env) GetBootImageName() string {
-	if "" != cToGoString(l.env.Bootimg_file_name[:]) {
-		return cToGoString(l.env.Bootimg_file_name[:])
+	fn := l.Get("bootimg_file_name")
+	if fn != "" {
+		return fn
 	}
 	return BOOTIMG_DEFAULT_NAME
 }
+
```

Comment:
all these should take  *bootimgKernelMatrix, instead of operating on values I think. Also they probably should use names that relate to what they do on the k-v matrix more than the high-level op, also if MATRIX_ROW_KERNEL vs MATRIX_ROW_RECOVERY_SYSTEM and values would be passed in they could be reused, it seems, between run and recovery code even more

Reply:
yes I did this, it is much simpler all across the board now I think